pkgpanda/util.py

```diff
@@ -264,13 +259,36 @@ def make_file(name):
 
 
 def write_json(filename, data):
-    with open(filename, "w+") as f:
-        return json.dump(data, f, **json_prettyprint_args)
+    dumped_json = json_prettyprint(data=data)
+    write_string(filename, dumped_json)
 
 
 def write_string(filename, data):
-    with open(filename, "w+") as f:
-        return f.write(data)
+    """
+    Write a string to a file.
+    Overwrite any data in that file.
+
+    We use an atomic write practice of creating a temporary file and then
+    moving that temporary file to the given ``filename``. This prevents race
+    conditions such as the file being read by another process after it is
+    opened here but not yet written to.
+
+    It also prevents us from creating or truncating a file before we fail to
+    write data to it because of low disk space.
+    """
+    prefix = os.path.basename(filename)
+    tmp_file_dir = os.path.dirname(os.path.realpath(filename))
+    fd, temporary_filename = tempfile.mkstemp(prefix=prefix, dir=tmp_file_dir)
+    os.close(fd)
+
+    try:
+        with open(temporary_filename, 'w') as f:
+            f.write(data)
+        os.chmod(temporary_filename, 0o644)
```

Comment:
You should get the mode from the destination file, using `os.stat`, in case this function is ever used for private files.

Reply:
Good thinking - and keep this if the target file does not exist?

Reply:
OK, but please add a note to the doc string that if the file does not exist it will be created with permissions 0o644.